### PR TITLE
Create a proposal for RFC proposals

### PR DESCRIPTION
**Pull request by LorenzMeier**
This lays out an approach to agree on RFCs. @amilcarlucas @hamishwillee Feedback would be welcome.

### CODE REVIEW

**Comment by amilcarlucas** [comment]
I believe it is a step forward. That means that all PRs will get a feedback within two weeks ?

**Comment by LorenzMeier** [comment]
Yes, we just have to commit to a higher quality of service. Even if that means that the feedback is "too hard" or "you have to join the Dronecode dev call to discuss it, because we don't understand it".

**Comment by auturgy** [comment]
I’m not sure what this will achieve, beyond adding additional process - and hence time. There is a real risk it will be counterproductive.
If the core issue is responsiveness, a scheduled dev call (monthly?) may be all that is needed. I’d suggest either a specific mavlink call, or alternate between ArduPilot and DroneCode.
Perhaps the first one should be soon, to discuss this proposal.

**Comment by jlecoeur** [comment]
Thank you for proposing an RFC process. I think the main value will be:
> This will insure quality of the design as well as later inform developers who have not been part of the original decisions of the reasons

As this is an open source project, all developpers/contributors/users cannot to take part to every dev call, and finding design documents on the internet (forums/google docs/others) can be challenging. When discussion happen on a dev call, the conclusions should be added to the RFC discussion, which is great to keep track of the design process.

I think the proposal could be clearer regarding which contributions require an RFC. 
For example, the addition of messages and commands are probably fine with simple PRs. Is it the same for the addition of fields in messages when these fields can be safely ignored? However modifications of enums, removal/merging of messages, addition of transaction protocol probably require an RFC.

**Comment by dagar** [comment]
> I’m not sure what this will achieve, beyond adding additional process - and hence time. There is a real risk it will be counterproductive.

@auturgy I don't see that we have a choice for something like Mavlink where we can be stuck with things that need to be maintained indefinitely.

File: text/0001-rfc-format.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+  * Start date: 2018-02-21
+  * Contributors: Lorenz Meier <lorenz@px4.io>, ...
+  * Related issues: 
+  
+# Summary
+
+The scope of this document is to design a Request-for-comments process for MAVLink for protocol additions that require a non-trivial change. The goal is to create a transparent, agreed-on process that allows the community and maintainer to make informed decisions and to capture the rationales behind these decisions in a document. This will insure quality of the design as well as later inform developers who have not been part of the original decisions of the reasons. The documents should also include further references and examples.
```
Comment (hamishwillee) [inline text/0001-rfc-format.md]: insure -> ensure (this is arguable, but ensure is better).

File: text/0001-rfc-format.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  
+# Motivation
+
+As the amount of micro-services in the MAVLink protocol increases the design process needs more structure to be able to accomodate all stakeholders and create the required transparency.
+
+# Detailed Design
+
+The RFC process should follow a simple flow:
+
+  * The original proposer raises a pull request with a document in this format.
```
Comment (hamishwillee) [inline text/0001-rfc-format.md]: You mean the format of this document? So "The RFC proposer raises a pull request with the same format/headings as this proposal."

File: text/0001-rfc-format.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  * The maintainer reviews the document and notifies additional stakeholders.
+  * Contributors comment on the draft or raise a separate PR. The original proposer is expected to be responsive for a period of two weeks so the document can converge and the technical discussion settle.
+  * Once the discussion has converged (or if more than four weeks have passed) the RFC is either merged (and the original proposer is expected to deliver the corresponding implementation) or rejected
```
Comment (hamishwillee) [inline text/0001-rfc-format.md]: Terminate with "."

**Comment by hamishwillee** [comment]
FWIW I agree with https://github.com/mavlink/rfcs/pull/1#issuecomment-367590195
- It is good to have a standardised approach for collecting information about proposed change.
- Overhead is bad, but if you can't articulate a proposal using a format as simple as this, then you probably aren't ready to implement it. Even if you want dev calls to be the main discussion area, having a document like this to start with will inform people with detail before you get there.
- It is good to have all discussion on a particular issue happen in a timely manner
- It is good that the results of the discussion are captured in one place.

This will only work if there is a nominated "Maintainer"

**Comment by auturgy** [comment]
To clarify, I’m not opposed to having a structure for documenting system design, or informing the development cycle. I actually love such things (need->spec->develop->review->approve->accept etc).
I think there is also value in arguing the counterpoint though, to ensure second and third order effects are considered before implementation.

For a number of reasons Mavlink development needs to prioritise stability over responsiveness, and this may help that. This should also make it easier to sync the two main projects, as changes to master here would be forced to document consideration of things such as pymavlink, which lives in another repo (well upstream does, and the pip package is built from that).
I do think we need to be careful that this rfc process doesn’t become a mechanism to force decisions, kill ideas, or dictate timeframes - different projects and people work at different paces, for different reasons. And that isn’t (necessarily) a bad thing.
Mavlink2, as an example, wouldn’t have met the timeframes suggested in this ‘policy’.

**Comment by LorenzMeier** [comment]
I'll do a new PR accommodating the additional feedback (or do one and I'll merge). I wanted to have something basic in.

**Pull request merged**